### ISSUE TITLE
Profiling Libtorch code using autogrid

### ISSUE THREAD

**Issue by abhiutd** [pytorch/pytorch#18630]
## 📚 Documentation

I have a Pytorch C++ frontend (LibTorch) based deployment codebase. I am trying to add profiling support to it. I am thinking of using autograd profiler for it, which seems to be the best option as far as getting layer-by-layer timings is concerned.
I recently saw a change getting pushed in which enables access to the autograd profiler from the C++ frontend (https://github.com/pytorch/pytorch/pull/16580). Is there any documentation for its usage ? Or some sort of a primer...?

**Comment by soumith** [comment]
unfortunately, no  -- it's poorly documented on the C++ end.
However, you can look at the unit test for example usage: https://github.com/pytorch/pytorch/pull/16580/files

**Comment by abhiutd** [comment]
Yeah, I am using `autograd::profiler::RecordProfile guard(filename)` to generate a trace file for now. I had a few queries regarding it ---

1. It seems to be a generating a list of events like
`{
  "name": "_convolution",
  "ph": "X",
  "ts": 20555.611000,
  "dur": 1865.102000,
  "tid": 0,
  "pid": "CPU Functions",
  "args": {}
},`

"name" - operator, 
"ts" - time stamp when the op started, 
"dur" - duration of op run. 
"pid" - hardcoded to CPU Functions 

(a) What is "ph" ?
(b) Why is "pid" CPU Functions even if I run my code on a GPU...?
(c) To generate a profile sequence, I can just arrange them based on the time stamps and duration ?

2. Whenever I am trying to change `ProfilerState` to lets say NVTX from CPU, I am getting an error - `can't change kind of profiling (e.g. NVTX to CPU) while profiler is running` . I have tried before loading the jit model, outside the scope of `RecordProfile` etc. 

(a) Am I missing something ?
(b) Would changing `ProfilerState` to `NVTX` give me something different as compared to the above event list ?

**Comment by labor00** [comment]
The profiler is generating a json file to be  visualized using the google chrome (chrome://tracing/).
This format is documented here (https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview). Due to the format convention is common to people to assign different meanings for things that in the original use case would be a pid. For instance in the horovod timeline the pid field is used to identify the tensor name, being reduced.

**Comment by iammarvelous** [comment]
Hi @soumith, I'm trying to profile a model with libtorch inside ios. It seems that profiler is not able to be linked. I got the below linking error when compile the app. My libtorch (1.10.0) is installed with cocopads. Do I need to custom build libtorch for ios? How do I include the profiler into the lib? Any pointer will be appreciated.
```
Undefined symbol: torch::autograd::profiler::RecordProfile::RecordProfile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
```
